src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp

```diff
@@ -0,0 +1,935 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <pup.h>
+
+#include "DataStructures/CachedTempBuffer.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/AnalyticSolution.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Solutions.hpp"
+#include "PointwiseFunctions/GeneralRelativity/TagsDeclarations.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace Tags {
+template <typename Tag>
+struct dt;
+}  // namespace Tags
+namespace gsl {
+template <class T>
+class not_null;
+}  // namespace gsl
+/// \endcond
+
+namespace gr {
+namespace Solutions {
+/*!
+ * \brief Schwarzschild black hole in Cartesian coordinates with harmonic gauge
```

Comment:
Standard Schwarzschild coordinates are not harmonic.  The simplest transformation to get a harmonic coordinate system from standard Schwarzschild coordinates is `r_H = r_S - M`.  Harmonic coordinates are also known as de Donder coordinates.  Goes back at least to a paper by Lanczos (1922).

Reply:
@macedo22 's implementation is just based on the implementation in SpEC, which @kidder did 16 years ago. Presumably it just used the transformation you mentioned. @kidder ? The code itself has no documentation explaining how the form of Schwarzschild was arrived at: https://github.com/sxs-collaboration/spec/blame/ad7fb23be8de1a725b38bef92c5a76de13524a97/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/HarmonicSchwarzschild.hpp

Reply:
@nilsvu @geoffrey4444 I've gone ahead and addressed almost all the change requests in the fixup I just pushed. The remaining tasks: documenting a literature reference (not sure what is best to cite/reference, please let me know!) and getting all of the new harmonic condition tests to pass (more on this below).

Regarding testing that the implementation satisfies eq 4.42, 4.44, and 4.45 (Baumgarte & Shapiro), it seems that only 4.44 holds (I've marked 4.42 and 4.45 with `TODO` comments in the test file in the fixup). What's strange is that I also copied over the SpEC implementation (and tweaked syntax so it compiles), and this PR's implementation matches the previous SpEC implementation's results in an additional test case I included in the fixup, so I'm not sure why 4.42 and 4.45 are not also satisfied.

Am I perhaps setting up the test in a weird way (e.g. weird center / bounds / coords / mass)?

Reply:
Thanks for checking so thoroughly. I believe what we have here are only _time-harmonic_ slices of Schwarzschild, Eq. 4.44. The spatial coords are not harmonic, Eq. 4.45. You actually copied a comment from SpEC in the test below that confirms this. So I suggest you add this information to the docs and remove the failing checks for Eq. 4.42 and 4.45.

Reply:
Thank you!

Reply:
With your recent corrections (missing square) it looks like your implementation differs from Eq. (45)--(50) in https://journals.aps.org/prd/abstract/10.1103/PhysRevD.56.4775 only by `r -> M + r` (I haven't checked this thoroughly). @geoffrey4444 and @markscheel what do you make of that?
- Should this class implement Eq. (45)--(50) exactly instead?
- Any ideas why SpEC has the additional `+M`?
- Is my assumption correct that Eq. (45)--(50) are the _time and space_ harmonic Kerr slices presented earlier in the paper, in the limit of zero spin? (I don't see trivially how Eq. (22) becomes Eq. (45), but I haven't gone through the math.)

@markscheel Please add the information we gather here to the docs.

Reply:
@macedo22: @markscheel and I discussed this and looked at the paper + the SpEC implementation. Our hypothesis is this:

- SpEC has a `TimeHarmonicSchwarzschild` solution that is _only_ time harmonic. This is Eq. (45)--(50) in the 1997 paper.
- What SpEC calls `HarmonicSchwarzschild`, and what you have implemented, is time _and_ spatially harmonic. It differs from `TimeHarmonicSchwarzschild`, and hence from the paper, by `r - > M + r`. It is the zero-spin limit of the `HarmonicKerr` solution in SpEC.

We want time _and_ spatially harmonic coordinates so Lorentz boosts preserve the harmonic time slicing. If our assumptions above are correct, then this is what you have implemented and everything is fine. Our suggestion is therefore:

1. Please add this information to the docs.
2. Please check again that the spatial coordinates are also harmonic, i.e., Eq. 4.45 in Baumgarte/Shapiro (the one that you said failed earlier). If this still fails then there's something we don't understand..

Reply:
I checked again and as before, eq 4.42 and 4.45 fail even though the unit tests shows that this spectre implementation matches the spec imlpementation result. It's definitely possible I have made a mistake somewhere, so I will keep looking into it, but I just wanted to let you know these conditions fail even after having already looked at it for a moderate amount of time and adding things like the spec impl test

Reply:
Yep, just fixed a mistake I made for 4.42 (I made a bad assumption about the equation)

Reply:
@nilsvu @geoffrey4444 In the most recent fixup (0fa1e008d), I fixed my test for 4.42 and got it to pass. However, 4.45 still won't pass. I've tried the following things:
- Added the spec implementation and tested against the spectre implementation --> the unit test shows they are equivalent
- Rewrote the 4.45 equation in python to make sure it wasn't a bug in `TensorExpression`s ---> same failing result
- Reimplemented the whole Harmonic Schwarzschild solution in Mathematica and checked 4.45 as well ---> nearly the same failing result (some numerical error probably)
- Instead of calculating the derivative of the natural log of the lapse analytically as I had previously done, the most recent fixup uses a numerical derivative instead to make sure the condition failing wasn't due to me making a mistake in the analytical expression I came up with on paper ---> using the numerical derivative didn't fix the issue

After these attempts, I'm not sure what else I can try, so I was wondering if maybe it would be good for a second pair of eyes to take a close look at my unit test and make sure I'm not doing something weird, like setting it up with strange parameters, or misinterpreting eq 4.45. If not, let me know what else I can try!

Reply:
That you got Eq. 4.42 means that the coordinates are indeed harmonic in space and time, so that's good! It also means that both Eq. 4.44 and 4.45 should pass, since they are just the 3+1 version of Eq. 4.42.

In my comment below I suggest you get rid of the numerical derivative, but from your comment here I understand that you have tried that already (still do it please). So I went ahead and derived Eq. 4.45 from Eq. 4.42 and got almost the result printed in Baumgarte/Shapiro, except for a minus sign. That might well be my mistake, but it might also be a typo in the book, so please check:

```tex
% Eq. 4.45 (check sign of \Gamma contraction term)
(\partial_t - \beta^j \partial_j) \beta^i
  = -\alpha^2 (\gamma^{ij} \frac{\partial_j \alpha}{\alpha} - \gamma^{jk} \Gamma^i_{jk})
```

You can also squash all fixup commits together to clean up this PR a bit, and add the information we gathered here to the docs.

Reply:
Oh my gosh @nilsvu the book did just have the sign incorrect. Switching it makes the test pass! I was starting to feel crazy. Thanks so much! I'll update things as soon as I can

Reply:
Nice!

Reply:
@nilsvu I've gone ahead and squashed the previous fixups, rebased, and added two new fixups so that you can see the new changes:
- first fixup: updates the test file to use the anlytical derivative and fixes the 4.45 check
- second fixup: documenting the result of our discussion on the equations in this thread (unsure if I described this correctly)

Reply:
3rd fixup: removing unneeded includes after removing numerical derivative